### DISCUSSION TITLE
theia-cloud for azure kubernetes service (aks)

### DISCUSSION THREAD

**Discussion by hkuich** [eclipse-theia/theia-cloud#229] [Q&A]
dear all,

i am working to translate the gke_getting_started to aks and feel like I'm about 80% there. AKS does a few things differently with their cluster configuration, and I wanted to see if there are other people who'd be willing and interested to contribute. I have opened a PR (https://github.com/eclipsesource/theia-cloud/pull/228) and am more than happy to contribute and then maintain the template once finished. We will be using this in production sooner rather than later and I will volunteer to maintain it for as long as we are active on azure.

my current issue is that aks creates two public IPs instead of one for the load balancer (it might create two load balancers, actually), and that confuses keycloak & theia because keycloak is hiding behind IP2, while theia is expecting it to be behind the IP1 (the initial one attached to the ingress load balancer). Any help would be greatly appreciated!

**Comment by hkuich** [comment]
I have also followed the azure documentation:

https://github.com/eclipsesource/theia-cloud/blob/azure-documentation/doc/docs/platforms/Azure.md

setting up a cluster this way works, and I see that there is the manual step of retrieving and setting the Cluster IP for the Keycloak and Theia-Cloud helm charts... I will try to and put that into terraform logic next.

However, after creating a realm and user etc, I run into a CORS issue. After I go to the landing page and log in with the user I created in keycloak, theia finally delivers a 401 and the browser console states:

`Access to XMLHttpRequest at 'https://service.MYIP.nip.io/service' from origin 'https://theia.cloud.MYIP.nip.io' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.`

Do you have any idea where I have to enable this in the setup?

**Comment by jfaltermeier** [comment]
Hi, thank you, this sounds really good. 
At the moment we don't have an Azure Account for testing, so I can't directly contribute atm. But below are my initial thoughts after looking at the PR:

I think these are the locations where we set the hostname for keycloak used by TheiaCloud:
https://github.com/verisianHQ/theia-cloud/blob/205af3ab56cdae1330a2bf36e7871813522f6f72/terraform/configurations/aks_getting_started/aks_getting_started.tf#L76
https://github.com/verisianHQ/theia-cloud/blob/205af3ab56cdae1330a2bf36e7871813522f6f72/terraform/modules/helm/main.tf#L235
So we have to make sure that those are right. 

To get the hostname on GKE we create an IP upfront that will be used by the nginx ingress via:
https://github.com/verisianHQ/theia-cloud/blob/205af3ab56cdae1330a2bf36e7871813522f6f72/terraform/configurations/gke_getting_started/gke_getting_started.tf#L42
https://github.com/verisianHQ/theia-cloud/blob/205af3ab56cdae1330a2bf36e7871813522f6f72/terraform/modules/helm/main.tf#L122

So I guess we could do something similar on Azure as well. 
https://learn.microsoft.com/en-us/azure/aks/ingress-tls?tabs=azure-cli#ingress-controller-configuration-options
In the creation chart I can see that we access public IPs already: https://github.com/verisianHQ/theia-cloud/blob/205af3ab56cdae1330a2bf36e7871813522f6f72/terraform/modules/cluster_creation/aks/main.tf#L65
But I think we haven't created those explicitly, have we? So maybe those IPs are all bound to something already and therefore the nginx-controller gets new one?
Maybe https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/public_ip.html helps

Regarding CORS:
Can you check the valid redirect uri used for setting up the keycloak realm? https://github.com/verisianHQ/theia-cloud/blob/205af3ab56cdae1330a2bf36e7871813522f6f72/terraform/modules/keycloak/main.tf#L16
https://github.com/verisianHQ/theia-cloud/blob/205af3ab56cdae1330a2bf36e7871813522f6f72/terraform/configurations/aks_getting_started/aks_getting_started.tf#L88
Maybe there is a hickup with the IPs as well.

**Reply by hkuich** [reply]
@jfaltermeier thank you so much for your extensive response! I have had to take a quick break on this - but we have someone joining who will be back on this by next week.
